aiohomekit/controller/coap/connection.py

```diff
@@ -0,0 +1,830 @@
+                return None
+
+        logger.debug('Paired with CoAP HAP accessory at %s!' % (self.address,))
+        await self.pairSetupClient.shutdown()
+        self.pairSetupClient = None
+
+        return pairing
+
+    async def do_pair_verify(self, pairing_data):
+        root = resource.Site()
+        coapClient = await Context.create_server_context(root, bind=('::',0))
+        uri = "coap://%s/2" % (self.address)
+
+        state_machine = get_session_keys(pairing_data)
+
+        request, expected = state_machine.send(None)
+        while True:
+            try:
+                payload = TLV.encode_list(request)
+                request = Message(code=Code.POST, payload=payload, uri=uri)
+                response = await coapClient.request(request).response
+                payload = TLV.decode_bytes(response.payload, expected=expected)
+
+                request, expected = state_machine.send(payload)
+            except StopIteration as result:
+                derive = result.value
+                break
+            except:
+                logger.warn('Pair verify failed!')
+                return None
+
+        recvKey = derive('Control-Salt', 'Control-Read-Encryption-Key')
+        recvCtx = ChaCha20Poly1305(recvKey)
+        sendKey = derive('Control-Salt', 'Control-Write-Encryption-Key')
+        sendCtx = ChaCha20Poly1305(sendKey)
+        eventKey = derive('Event-Salt', 'Event-Read-Encryption-Key')
+        eventCtx = ChaCha20Poly1305(eventKey)
+        self.encCtx = EncryptionContext(recvCtx, sendCtx, eventCtx)
+
+        logger.debug('Connected to CoAP HAP accessory at %s!' % (self.address,))
+        self.connection = coapClient
+        root.add_resource([], EventResource(self))
+
+        return True
+
+    async def connect(self, pairing_data):
+        async with self.connectionLock:
+            if self.is_connected():
+                logger.debug('Already connected')
+                return True
+
+            await self.do_pair_verify(pairing_data)
+            await self.get_accessory_info()
+
+    def is_connected(self):
+        return self.connection != None and self.encCtx != None
+
+    async def _post(self, payload):
+        uri = "coap://%s/" % (self.address)
+
+        async with self.lock:
+            payload = self.encCtx.encrypt(payload)
+
+            try:
+                request = Message(code=Code.POST, payload=payload, uri=uri)
+                response = await self.connection.request(request).response
+            except NetworkError:
+                raise AccessoryDisconnectedError('Request timeout')
+
+            if response.code != Code.CHANGED:
+                logger.warn('CoAP POST returned unexpected code %s' % (response.code,))
+
+            try:
+                payload = self.encCtx.decrypt(response.payload)
+            except InvalidTag as ex:
+                logger.error('Decryption failed, desynchronized? Counter=%d/%d' % (self.encCtx.recvCtr, self.encCtx.sendCtr))
```

Comment:
Worth pulling this into `EncryptionContext`? Easier to test there as you don't have to mock any i/o, and avoids leaking `recvCtr` awareness into your connection code.

Reply:
I moved the post function to the encryption context class. I'll probably do a little more refactoring here. I don't like that the device address is retained in two different places.